Compute CHM using rioxarray

russelldj opened the pull request:
This roughly follows the logic [here](https://github.com/open-forest-observatory/ofo-catalog-data-prep/blob/c0a7733965a0528e67deaa084ca7e80fbbec52d6/deploy/drone-imagery-ingestion/03_photogrammetry/src/20_postprocess-photogrammetry-products.R#L43) but additionally adds the option to set a spatial resolution for the output CHM. Also, it threshold the values at 0, which I saw implemented in a different CHM method. Is the latter desirable?

Also, thanks @youngdjn for the suggestion to look into `rioxarray`. It seems quite powerful.

youngdjn commented on ground_reference_prep/1_compute_CHM.py:
```diff
@@ -0,0 +1,61 @@
+    dtm = rioxarray.open_rasterio(dtm_path, masked=True)
+    dsm = rioxarray.open_rasterio(dsm_path, masked=True)
+
+    if resolution is not None:
```
I think you need to remove the `not` or flip the code between `if` and `else`.

youngdjn replied:
Ah I see you fixed this as I was commenting!

youngdjn commented on ground_reference_prep/1_compute_CHM.py:
```diff
@@ -0,0 +1,61 @@
+        )
+
+    # Pixel-align the dsm to the DTM
+    dsm_resampled = dsm.rio.reproject_match(dtm_resampled)
```
I would prefer to align the DTM to the DSM instead (also requires changing the code above to operate on DSM instead of DTM) because there's a future possibility to use a third-party existing DTM, which might be in a very different resolution than the photogrammetry DSM. Also more generally (applies to the case of no resolution specified), I feel more comfortable resampling a DTM because it should have less detail than a DSM.

youngdjn reviewed (changes requested):
Looking good! I like the idea to set a floor of 0.

If the target resolution is many factors coarser than the input resolution, do you know how `rioxarray` deals with the resampling? Does it take a weighted average of every input cell that falls within each target cell, weighted by the fraction of the input cell falling within the target cell? Does it just take the most central input cell? I'm actually not sure how important this distinction is, but my intuition is we prefer the former.

russelldj commented:
> Looking good! I like the idea to set a floor of 0.
> 
> If the target resolution is many factors coarser than the input resolution, do you know how `rioxarray` deals with the resampling? Does it take a weighted average of every input cell that falls within each target cell, weighted by the fraction of the input cell falling within the target cell? Does it just take the most central input cell? I'm actually not sure how important this distinction is, but my intuition is we prefer the former.

I'll read up on how bilinear resampling works for coarstening and if there are any better options.

russelldj commented:
Good intuition here. Yes, it looks like aggressive coarsening with bilinear (or bicubic) interpolation runs into issues with aliasing for thin structures as described [here](https://stackoverflow.com/questions/49879466/downscaling-images-using-bilinear-and-bicubic-algorithms). I found that rasterio has an `average` resampling strategy, which seems like what you were suggesting. And is also consistent with the theory I've found for this. I did run both and saw little practical difference, probably because of the relatively smooth transitions in CHM values. Average is slightly faster than bilinear.

**full-res**
![image](https://github.com/user-attachments/assets/9515f39d-d433-4e3a-b7a1-4191cf62791d)
**bilinear**
![image](https://github.com/user-attachments/assets/03def1cd-c87d-4021-acda-34e2732a243f)
**average**
![image](https://github.com/user-attachments/assets/386d1aad-d42d-460c-aa64-e88090942567)

We could select average if the image is being downsampled and bicubic (seems slightly better than bilinear) if it's being upsampled.

russelldj commented:
I've addressed the comments you left and this works for paired photogrammetry DSMs and DTMs. However, I tried it with a USGS DTM and [this comment](https://github.com/open-forest-observatory/tree-species-prediction/blob/c93f19d42b598f6300131868479a8ccc0f9697b0/ground_reference_prep/1_compute_CHM.py#L55) became relevant. Since the DTM was in geographic coordinates, it seemed like it was higher resolution and it was upsampled with averaging, which effectively becomes nearest neighbor and leads to blocky artificats. I think there should be an efficient way (without actually reprojecting the underlying raster data) to determine the spatial resolution in meters with [calculate_default_transform](https://rasterio.readthedocs.io/en/latest/api/rasterio.warp.html#rasterio.warp.calculate_default_transform) or similar, but I'll have to work through the logic.

youngdjn reviewed (approved):
Great! Looks good to merge.

Good test, and good points about the geographic coordinates issue. It seems the central issue is needing to know the resolution of the (geographic-coords) DTM _in meters_. I don't know a good way other than doing a preliminary projection of the raster (potentially a cropped subset) and querying the res. However I do know there exist tools for doing meters-based operations on geographic data, specifically [S2 Geometry](https://github.com/google/s2geometry), but I don't know how to use it outside of its [sf bindings](https://r-spatial.github.io/sf/reference/s2.html) (e.g. for [st_buffer](https://r-spatial.github.io/sf/reference/geos_unary.html)). Probably not worth the time to implement here unless there are similar bindings in rasterio.

The z offset and edge artifact issues when using the USGS DEM make sense -- good to be aware of.
